### DISCUSSION TITLE
No way to properly escape single, multi and 'all' option variables?

### DISCUSSION THREAD

**Discussion by adam-nielsen** [grafana/grafana#43414] [General]
How can I properly escape an SQL query (in my case Postgres/TimescaleDB) such that it works when a variable has a single value selected, multiple values selected, and the 'All' value?

For example, if I have a variable called `test`, it does not seem possible to use it in a query like this:
```
...WHERE col IN ($test)
OR $test = 'All'
```
This works for single values, but if I set `$test` to multiple values, the query breaks:
```
...WHERE col IN ('one', 'two')
OR 'one', 'two' = 'All'
        ^ syntax error
```
I can use the `rawstring` formatter to fix this:
```
...WHERE col IN ($test)
OR '${test:rawstring}' = 'All'
```
Which comes out as:
```
...WHERE col IN ('one', 'two')
OR 'one,two' = 'All'
```
However if I set the variable to "all values" and specify a "custom all value" of `All` then it breaks:
```
...WHERE col IN (All)   <-- error, missing quotes
OR 'All' = 'All'
```
If I put quotes in the custom value so it's `'All'` then it fixes the `IN` part but breaks the `OR` part:
```
...WHERE col IN ('All')
OR ''All'' = 'All'
    ^ too many quotes
```
If I remove the quotes from the `OR` part then it works, but breaks again once I select a single value:
```
...WHERE col IN ($test)
OR ${test:rawstring} = 'All'

...WHERE col IN ('All')
OR 'All' = 'All'

...WHERE col IN ('one')
OR one = 'All'
   ^ error, missing quotes
```

In other words, I can't see a way to  use a variable for both an `IN` and `=` SQL condition, while also allowing a single value to be selected, multiple values to be selected, or a custom value for the All option.

Am I missing something or is this not currently possible?

**Comment by gabor** [comment]
hi, i created a github-issue based on the description, it's at https://github.com/grafana/grafana/discussions/43414
